### PR TITLE
feat: add new errorTypes (firehose, freshsales, freshmarketer, ga, ga4)

### PR DESCRIPTION
**Pull request by ujjwal-ab**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/freshmarketer/utils.js [hunk 1]
```diff
@@ -320,9 +341,8 @@ const UpdateContactWithLifeCycleStage = async (message, Config) => {
       lifeCycleStage => lifeCycleStage.name === lifecycleStageName
     );
     if (!lifeCycleStageDetials) {
-      throw new CustomError(
-        `failed to fetch lifeCycleStages with ${lifecycleStageName}`,
-        400
+      throw new InstrumentationError(
```
Comment (saikumarrs) [inline src/v0/destinations/freshmarketer/utils.js:344]: Should this be `NetworkInstrumentationError`?

File: src/v0/destinations/freshsales/utils.js [hunk 2]
```diff
@@ -326,9 +350,8 @@ const UpdateContactWithLifeCycleStage = async (message, Config) => {
       lifeCycleStage => lifeCycleStage.name === lifecycleStageName
     );
     if (!lifeCycleStageDetials) {
-      throw new CustomError(
-        `failed to fetch lifeCycleStages with ${lifecycleStageName}`,
-        400
+      throw new InstrumentationError(
```
Comment (saikumarrs) [inline src/v0/destinations/freshsales/utils.js:353]: `NetworkInstrumentationError`?

File: src/v0/destinations/ga/networkHandler.js [hunk 3]
```diff
@@ -26,16 +30,11 @@ const gaResponseHandler = gaResponse => {
       return errObj.reason && errObj.reason === "invalidCredentials";
     });
     if (isInvalidCredsError || response?.error?.status === "UNAUTHENTICATED") {
-      throw new ErrorBuilder()
-        .setMessage("[GA] invalid credentials")
-        .setStatus(500)
-        .setDestinationResponse(response)
-        .setAuthErrorCategory(REFRESH_TOKEN)
-        .build();
+      throw new RetryableError("invalid credentials");
```
Comment (saikumarrs) [inline src/v0/destinations/ga/networkHandler.js:33]: Auth error category and destination response should be set.

File: src/v0/destinations/ga/networkHandler.js [hunk 4]
```diff
@@ -26,16 +30,11 @@ const gaResponseHandler = gaResponse => {
...
     }
-    throw new ErrorBuilder()
+    throw new NetworkInstrumentationError()
```
Comment (saikumarrs) [inline src/v0/destinations/ga/networkHandler.js:35]: 1. This is not exactly an instrumentation error.
2. Previous error builder API is used to set other error fields.

File: src/v0/destinations/ga4/networkHandler.js [hunk 5]
```diff
@@ -38,39 +43,21 @@ const responseHandler = (destinationResponse, dest) => {
...
-        .isTransformResponseFailure(true)
-        .setDestinationResponse(response?.validationMessages[0]?.description)
-        .setStatTags({
-          destType: dest,
-          stage: TRANSFORMER_METRIC.TRANSFORMER_STAGE.RESPONSE_TRANSFORM,
-          scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.SCOPE,
-          meta: getDynamicErrorType(status)
-        })
-        .build();
+      throw new NetworkInstrumentationError(
```
Comment (saikumarrs) [inline src/v0/destinations/ga4/networkHandler.js:46]: This should be `NetworkError`. Also, the error type and other fields need to be set.

File: src/v0/destinations/ga4/networkHandler.js [hunk 6]
```diff
@@ -38,39 +43,21 @@ const responseHandler = (destinationResponse, dest) => {
...
-      .isTransformResponseFailure(true)
-      .setDestinationResponse(destinationResponse)
-      .setStatTags({
-        destType: dest,
-        stage: TRANSFORMER_METRIC.TRANSFORMER_STAGE.RESPONSE_TRANSFORM,
-        scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.SCOPE,
-        meta: getDynamicErrorType(status)
-      })
-      .build();
+    throw new NetworkError(
```
Comment (saikumarrs) [inline src/v0/destinations/ga4/networkHandler.js:54]: Destination response is not set.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: src/v0/destinations/ga/networkHandler.js (hunk 4 above)
Comment (utsabc) [reply src/v0/destinations/ga/networkHandler.js:35]: This is a NetworkError

**Review by utsabc — APPROVED** [review]

File: src/v0/destinations/ga/deleteUsers.js [hunk 7]
```diff
@@ -17,23 +21,17 @@ const prepareDeleteRequest = (userAttributes, config, rudderDestInfo) => {
...
-      .setMessage(
-        // This would happen when server doesn't send "x-rudder-dest-info" header
-        // Todo's in-case this exception happen:
-        // 1. The server version might be an older one
-        // 2. There would have been some problem with how we are sending this header
-        `The "secret" field is not sent in "x-rudder-dest-info" header`
-      )
-      .setStatus(500)
-      .build();
+    throw new RetryableError(
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply src/v0/destinations/ga/deleteUsers.js:24]: Yeah, this should be `OAuthSecretError`.

File: src/v0/destinations/ga/networkHandler.js [hunk 8]
```diff
@@ -26,20 +47,21 @@ const gaResponseHandler = gaResponse => {
...
-        .setMessage("[GA] invalid credentials")
-        .setStatus(500)
-        .setDestinationResponse(response)
-        .setAuthErrorCategory(REFRESH_TOKEN)
-        .build();
+      throw new RetryableError(
+        "invalid credentials",
+        response?.error?.status,
+        response,
+        getAuthErrCategory(response?.error?.status)
```
Comment (utsabc) [inline src/v0/destinations/ga/networkHandler.js:54]:
```suggestion
        getAuthErrCategory(REFRESH_TOKEN)
```

File: src/v0/destinations/ga/networkHandler.js [hunk 9]
```diff
@@ -2,9 +2,30 @@ const {
...
+ */
+const getAuthErrCategory = (code, response) => {
+  switch (code) {
+    case 401:
+      if (!response.error?.details) return REFRESH_TOKEN;
+      return "";
+    default:
+      return "";
+  }
+};
```
Comment (utsabc) [inline src/v0/destinations/ga/networkHandler.js:28]: Lets not introduce drastic changes right now

Comment (saikumarrs) [reply src/v0/destinations/ga/networkHandler.js:28]: @ujjwal-ab where did this info come from?

Comment (ujjwal-ab) [reply src/v0/destinations/ga/networkHandler.js:28]: We are using this in many destinations to get `AuthErrCategory`.

Comment (utsabc) [reply src/v0/destinations/ga/networkHandler.js:28]: Lets not make any drastic changes

File: src/v0/destinations/ga4/networkHandler.js [hunk 10]
```diff
@@ -38,39 +40,27 @@ const responseHandler = (destinationResponse, dest) => {
...
-          meta: getDynamicErrorType(status)
-        })
-        .build();
+      throw new NetworkError(
+        `Validation Server Response Handler:: Validation Error for ${dest} of field path :${fieldPath} | ${validationCode}-${description}`,
+        400,
+        {
+          [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(status)
+        },
+        response?.validationMessages[0]?.description
```
Comment (saikumarrs) [inline src/v0/destinations/ga4/networkHandler.js:49]: Should it directly be `destinationResponse`?

Comment (ujjwal-ab) [reply src/v0/destinations/ga4/networkHandler.js:49]: Previously, in errorBuilder we were setting this only.

File: src/v0/destinations/ga/networkHandler.js (hunk 9 above)
Comment (saikumarrs) [reply src/v0/destinations/ga/networkHandler.js:28]: That's not how you are supposed to interpret the status code. Revert the changes for now and set the hardcoded auth err category like it was before.

File: src/v0/destinations/ga/networkHandler.js [hunk 11]
```diff
@@ -26,20 +29,21 @@ const gaResponseHandler = gaResponse => {
...
     if (isInvalidCredsError || response?.error?.status === "UNAUTHENTICATED") {
-      throw new ErrorBuilder()
-        .setMessage("[GA] invalid credentials")
-        .setStatus(500)
-        .setDestinationResponse(response)
-        .setAuthErrorCategory(REFRESH_TOKEN)
-        .build();
+      throw new RetryableError(
+        "invalid credentials",
+        response?.error?.status,
```
Comment (utsabc) [inline src/v0/destinations/ga/networkHandler.js:34]:
```suggestion
        500
```

File: src/v0/destinations/ga/networkHandler.js [hunk 12]
```diff
@@ -51,7 +33,7 @@ const gaResponseHandler = gaResponse => {
         "invalid credentials",
         response?.error?.status,
         response,
-        getAuthErrCategory(response?.error?.status)
+        "REFRESH_TOKEN"
```
Comment (saikumarrs) [inline src/v0/destinations/ga/networkHandler.js:36]: I meant to use the const we are importing above.

**Review by utsabc — APPROVED** [review]

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**